### PR TITLE
tests: add check for playbook execution

### PR DESCRIPTION
**Pull request by sometheycallme**
#104

### CODE REVIEW

**Comment by kropped** [comment]
:bell: Staged `cleanerbot_tests` at revision 4bd905993d7b318295929346314361e59669a453 on 8ea6abddb6d7

**Comment by sometheycallme** [comment]
for manually testing in CircleCI the helpful_files/env_vars are important to the docker run commands in shell scripts.

Trying to catch this line output:

example
```shell
@test "invoking foo without arguments prints usage" {
  run foo
  [ "$status" -eq 1 ]
  [ "${lines[0]}" = "usage: foo <filename>" ]
}
```

Circle still doesn't like it

< PLAY [local] *******************************************************************

**Comment by sometheycallme** [comment]
@jumanjiman 

To actually check for playbook execution, we need to access the output of the webserver response (one of the lines[])

Unfortunately similar to prior attempts circle only likes the first line output check.   

As a result the test is incomplete.  You really don't know if playbook runs

**Comment by jumanjiman** [comment]
circle mixes stderr and stdout, so you need to remove the index from lines var and just match any of the lines. meh.

**Comment by sometheycallme** [comment]
>circle mixes stderr and stdout, so you need to remove the index from lines var and just match any of the lines. meh.

I'll give it a try.

**Comment by sometheycallme** [comment]
[[ ]] allows for partial matching.  

I'll try [[ {$lines[0] =~ PLAY ]]

**Comment by kropped** [comment]
:bell: Staged `cleanerbot_tests` at revision a95b8f238b29c234937fa7da9a78581963c9e06d on 12e49fe2d675

**Comment by kropped** [comment]
:bell: Staged `cleanerbot_tests` at revision a95b8f238b29c234937fa7da9a78581963c9e06d on 31d1f257548a

**Comment by sometheycallme** [comment]
Okay, I manually spawned the pipeline for autostager, staging-data, and ansible-controller

After fixing the env_vars path in my shell scripts, the CURL test yields the following output:


```shell
ubuntu@box359:~$ curl -v -X POST -d '{"branch_name": "tests", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "fixtures/etc/ansible/play_test.yml"}' http://172.17.0.14:8080/play
* About to connect() to 172.17.0.14 port 8080 (#0)
*   Trying 172.17.0.14... connected
> POST /play HTTP/1.1
> User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
> Host: 172.17.0.14:8080
> Accept: */*
> Content-Length: 150
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 150out of 150 bytes
* HTTP 1.0, assume close after body
< HTTP/1.0 200 
< PLAY [local] *******************************************************************
< 
TASK [setup] *******************************************************************
ok: [localhost]

TASK [Ansible Ping Localhost] **************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   


Server: BaseHTTP/0.3 Python/2.7.12
Date: Mon, 11 Jul 2016 21:29:12 GMT
* Closing connection #0
```

**Comment by sometheycallme** [comment]
I've tried to look for ```PLAY```

File: script/test_ansible.bats [hunk 1]
```diff
@@ -30,10 +30,6 @@ load options
    ip=$(echo ${DOCKER_HOST} | awk -F/ '{print $NF}' | cut -d: -f0)
    port=$(docker port hooktest | awk -F: '{print $NF}')
 fi
- run curl -v -X POST -d '{"branch_name": "master", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "playbooks/git-clonerepos.yml"}' http://${ip}:${port}/play
-  
-  [[ ${output} =~ About ]]
+ run curl -v -X POST -d '{"branch_name": "master", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "fixtures/etc/ansible/play_test.yml"}' http://${ip}:${port}/play
+  [[ ${lines} =~ PLAY ]]
```
Comment (jumanjiman) [inline script/test_ansible.bats]: `${lines}` is an array.
try `[[ $output =~ PLAY ]]` instead.

**Comment by sometheycallme** [comment]
>try [[ $output =~ PLAY ]] instead.

Yes, and no likey, likey.

I'm going to see if I can pipe this into grep and get the desired output.

**Comment by sometheycallme** [comment]
https://github.com/sstephenson/bats/issues/10

**Comment by sometheycallme** [comment]
I just realized that in order to get the fixtures working (even if output was okay) we need to make sure ansibe-security fixtures is available inside the controller.  

This testing does not mimic the pipeline I've been building in DO, rather it tests each individual component.

**Comment by sometheycallme** [comment]
https://github.com/cleanerbot/ansible-security/commit/c9b7470e41fe0026ea5fbd00d925290696b48960

**Comment by sometheycallme** [comment]
Okay, getting closer.  Need to ensure the path to fixtures is set correctly:


```
Listening on localhost:8080


{u'branch_name': u'master', u'git_handle': u'cleanerbot', u'flags': [{u'flag': u'-i', u'argument': u'hosts'}], u'playbook': u'fixtures/etc/ansible/play_test.yml'}
fixtures/etc/ansible/play_test.yml
[{u'flag': u'-i', u'argument': u'hosts'}]
{u'flag': u'-i', u'argument': u'hosts'}

ERROR! the playbook: /opt/staging/cleanerbot_master/fixtures/etc/ansible/play_test.yml could not be found

program output: 
ansible-playbook -i hosts /opt/staging/cleanerbot_master/fixtures/etc/ansible/play_test.yml
/opt/staging/cleanerbot_master
172.17.42.1 - - [13/Jul/2016 14:10:52] "POST /play HTTP/1.1" 200 -
```

**Comment by sometheycallme** [comment]
http://stackoverflow.com/questions/30970969/exec-is-not-supported-by-the-lxc-driver-how-to-get-around-this

**Comment by sometheycallme** [comment]
https://circleci.com/docs/docker/#docker-exec

**Comment by sometheycallme** [comment]
The LXC to exec into the container is a pain in the ass.

stopped container and ran with entrypoint bash.

need to make sure that the fixtures data is actually copied in.  Directory is there and empty.

```
bash-4.3# pwd
/opt/staging/cleanerbot_master/ansible-security
bash-4.3# ls
bash-4.3# 
```

**Comment by sometheycallme** [comment]
@Jfach 

Here is the PR I've been hacking away at. 

There are two oustanding issues I'm trying to resolve.

1) When running the docker images for test_ansible.bats, mount the fixtures directory and check for the static playbook.

```shell
@test "ansible-controller: fixtures path is set" {
 # check to see if the $FIXTURES_DATA_IMAGE is properly set iun path for controller to see playbook
 run docker run -d --name=voltest -p 8080:8080 -v /fixtures/etc/ansible:/opt/staging/cleanerbot_master/ansible-security/ ansible-controller --entrypoint bash $CONTROLLER_IMAGE -c "ls -l /opt/staging/cleanerbot_master/ansible-security"
 [[ ${output} =~ play_test ]]
}
```

2) Check for output of the playbook running properly.

```shell
@test "ansible-controller: webserver responds to curl" {
 # run docker run -d --name=webtest -p 8080:8080 --volumes-from $FIXTURES_DATA_IMAGE:ro ansible-controller
 run docker run -d --name=webtest -p 8080:8080 -v /fixtures/etc/ansible:/opt/staging/cleanerbot_master/ansible-security/ ansible-controller
 if [[ x$DOCKER_HOST = x ]]; then
 # use local network namespace
   ip=$(docker inspect --format '{{ .NetworkSettings.IPAddress }}' webtest)
   port=8080
 else
 # accomodate remote docker execution.
   ip=$(echo ${DOCKER_HOST} | awk -F/ '{print $NF}' | cut -d: -f0)
   port=$(docker port hooktest | awk -F: '{print $NF}')
fi
 run curl -v -X POST -d '{"branch_name": "master", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "fixtures/etc/ansible/play_test.yml"}' http://${ip}:${port}/play
  # we curl and run 'ansible-playbook' against /opt/staging/cleanerbot_master/ansible-security/fixtures/etc/ansible/play_test.yml
  # the full path is seeded by webserver.py with `ANSIBLE_HOSTS` + git_handle + branch_name + path
  [[ ${output} =~ About ]]
}
```

**Comment by sometheycallme** [comment]
I am doing it wrong:

The absolute path of the ansible-security repository on the ubuntu system in circleCI needs to be provded like this:

```
ubuntu@box233:~/ansible-security$ docker run -i -t --name=voltest -p 8080:8080 --env-file helpful_files/env_vars -v /home/ubuntu/ansible-security/fixtures/etc/ansible:/opt/staging/cleanerbot_master/ansible-security --entrypoint bash ansible-controller 
bash-4.3# cd /opt/staging/
bash-4.3# ls
cleanerbot_master
bash-4.3# 
bash-4.3# cd cleanerbot_master/ansible-security/
bash-4.3# ls
group_vars     inventory      play_test.yml  site.yml
bash-4.3# pwd
/opt/staging/cleanerbot_master/ansible-security
bash-4.3# 

```

**Comment by sometheycallme** [comment]
Okay, better.  Now the fixtures playbook is in the correct path.

```
bats script/test_*.bats
1..5
ok 1 ansible-controller: Ansible 2.x is installed and running
ok 2 ansible-controller: Autostager staging directory for contributor branches is in path
ok 3 autostager: Autostager python script is in path
ok 4 ansible-controller: fixtures path is set
not ok 5 ansible-controller: webserver responds to curl
# (in test file script/test_ansible.bats, line 43)
#   `[[ ${output} =~ PLAY ]]' failed
make: *** [test] Error 1

make test returned exit code 2
```

**Comment by sometheycallme** [comment]
I am going to change 

```[[ ${output} =~ PLAY ]]```

to 

```[[ ${output} =~ About ]]```

And this test will pass.   

So either the playbook isn't running and in the output, or the syntax ```[[ ${output} =~ PLAY]]``` isn't respected.  My understanding is that this syntax allows for more flexible matching of output.

**Comment by sometheycallme** [comment]
Looks like the playbook doesn't execute:

```
ubuntu@box124:~/ansible-security$ curl -v -X POST -d '{"branch_name": "master", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "fixtures/etc/ansible/play_test.yml"}' http://172.17.0.4:8080/play
* About to connect() to 172.17.0.4 port 8080 (#0)
*   Trying 172.17.0.4... connected
> POST /play HTTP/1.1
> User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
> Host: 172.17.0.4:8080
> Accept: */*
> Content-Length: 151
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 151out of 151 bytes
* HTTP 1.0, assume close after body
< HTTP/1.0 200 
< Server: BaseHTTP/0.3 Python/2.7.12
< Date: Wed, 13 Jul 2016 18:16:44 GMT
* Closing connection #0
```

so let's see why:

```
ubuntu@box124:~/ansible-security$ docker run -i -t --name=webtest -p 8080:8080 --env-file helpful_files/env_vars -v /home/ubuntu/ansible-security/fixtures/etc/ansible:/opt/staging/cleanerbot_master/ansible-security ansible-controller bash
Listening on localhost:8080

```

**Comment by sometheycallme** [comment]
interesting:

```
{u'branch_name': u'master', u'git_handle': u'cleanerbot', u'flags': [{u'flag': u'-i', u'argument': u'hosts'}], u'playbook': u'/opt/staging/cleanerbot_master/ansible-security/fixtures/etc/ansible/play_test.yml'}
/opt/staging/cleanerbot_master/ansible-security/fixtures/etc/ansible/play_test.yml
[{u'flag': u'-i', u'argument': u'hosts'}]
{u'flag': u'-i', u'argument': u'hosts'}
ERROR! the playbook: /opt/staging/cleanerbot_master//opt/staging/cleanerbot_master/ansible-security/fixtures/etc/ansible/play_test.yml could not be found
program output: 
ansible-playbook -i hosts /opt/staging/cleanerbot_master//opt/staging/cleanerbot_master/ansible-security/fixtures/etc/ansible/play_test.yml
/opt/staging/cleanerbot_master
172.17.42.1 - - [13/Jul/2016 18:23:47] "POST /play HTTP/1.1" 200 -
```

**Comment by sometheycallme** [comment]
wrong path for playbook specified.
Should be
```
bash-4.3# ls 
group_vars     inventory      play_test.yml  site.yml
bash-4.3# pwd
/opt/staging/cleanerbot_master/ansible-security
bash-4.3# 
```

**Comment by sometheycallme** [comment]
hmm.  still no like.

ubuntu@box124:~/ansible-security$ curl -v -X POST -d '{"branch_name": "master", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "ansible-security/play_test.yml"}' http://172.17.0.12:8080/play
* About to connect() to 172.17.0.12 port 8080 (#0)
*   Trying 172.17.0.12... connected
> POST /play HTTP/1.1
> User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
> Host: 172.17.0.12:8080
> Accept: */*
> Content-Length: 147
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 147out of 147 bytes
* HTTP 1.0, assume close after body
< HTTP/1.0 200 
< PLAY [local] *******************************************************************
< skipping: no hosts matched
< 
PLAY RECAP *********************************************************************


Server: BaseHTTP/0.3 Python/2.7.12
Date: Wed, 13 Jul 2016 18:28:59 GMT
* Closing connection #0
```

**Comment by sometheycallme** [comment]
So I just added a test for checking the playbook execution locally within ansible, and with fixtures.

The test for anything in the first line of sdout seems to be okay.

I'm going to try now, testing for anything in the 2nd line of output with the ${lines[1]} syntax.

```
ubuntu@box124:~/ansible-security$ curl -v -X POST -d '{"branch_name": "master", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "ansible-security/play_test.yml"}' http://172.17.0.12:8080/play
* About to connect() to 172.17.0.12 port 8080 (#0)
*   Trying 172.17.0.12... connected
> POST /play HTTP/1.1
> User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
> Host: 172.17.0.12:8080
> Accept: */*
> Content-Length: 147
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 147out of 147 bytes
* HTTP 1.0, assume close after body
< HTTP/1.0 200 
< PLAY [local] *******************************************************************
< skipping: no hosts matched
< 
PLAY RECAP *********************************************************************


Server: BaseHTTP/0.3 Python/2.7.12
Date: Wed, 13 Jul 2016 18:28:59 GMT
* Closing connection #0
```

**Comment by sometheycallme** [comment]
A comment in the first example of README.md would be helpful. IMHO that's good enough, no need to have bats detect the use of double-brackets and issue a warning.

I've also been bitten by this, used double-brackets to test the output against a regex, like so:

run some cmd
[ "$status" -eq 0 ]
[[ "$output" =~ ^some\ +(random )?\ *regex$ ]]
The test command (i.e. [) doesn't support matching against regular expressions, so I resorted to using grep with a Here String instead:

run some cmd
[ "$status" -eq 0 ]
grep -E '^some +(random )? *regex$' <<< "$output"
Hope this helps others who stumble upon this issue.

Summary:

use [ instead of [[ where possible;
use grep with a Here String grep -E '...' <<< "$output" instead of double-bracket regex matching [[ "$output" =~ ... ]];
append || false if you still prefer to use [[, like so: [[ "$output" =~ blah ]] || false – though I find that quite error prone.

**Comment by sometheycallme** [comment]
Interesting messing with the output here:

```
not ok 6 ansible-controller: webserver responds to curl and playbook executes remotely (outside container)
# (in test file script/test_ansible.bats, line 42)
#   `run echo $testoutput | grep PLAY' failed
# * About to connect() to 172.17.0.8 port 8080 (#0)
# *   Trying 172.17.0.8...   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0connected
# > POST /play HTTP/1.1

# > User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3

# > Host: 172.17.0.8:8080

# > Accept: */*

# > Content-Length: 151

# > Content-Type: application/x-www-form-urlencoded

# > 

# } [data not shown]
# * upload completely sent off: 151out of 151 bytes
# * HTTP 1.0, assume close after body
# < HTTP/1.0 200 
# < PLAY [local] *******************************************************************
# < skipping: no hosts matched
# < 
# { [data not shown]
100   308    0   157  100   151    280    269 --:--:-- --:--:-- --:--:--   279
# * Closing connection #0
make: *** [test] Error 1
```

**Comment by sometheycallme** [comment]
@jumanjiman 

works but WOW that was ugly.

```shell
@test "ansible-controller: webserver responds to curl and playbook executes remotely (outside container)" {
 # check to see if ansible webserver accepts json data (commands) and runs fixtures playbook remotely
 # sending feedback of play execution to user
 run docker run -d --name=webtest -p 8080:8080 --env-file helpful_files/env_vars -v /home/ubuntu/ansible-security/fixtures/etc/ansible:/opt/staging/cleanerbot_master/ansible-security ansible-controller
   ip=$(docker inspect --format '{{ .NetworkSettings.IPAddress }}' webtest)
   port=8080
   sleep 5
   testoutput=$(curl -v -X POST -d '{"branch_name": "master", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "inventory"}], "playbook": "ansible-security/play_test.yml"}' http://${ip}:${port}/play)
   echo $testoutput > testfile
  run grep PLAY testfile
 [[ ${output} =~ PLAY ]]
}
```

**Pull request merged**